packages/components/src/base-control/stories/index.tsx

```diff
@@ -1,81 +1,79 @@
+/**
+ * External dependencies
+ */
+import type { ComponentMeta, ComponentStory } from '@storybook/react';
+
 /**
  * Internal dependencies
  */
-import BaseControl from '../';
+import BaseControl from '..';
 import Button from '../../button';
-import { Spacer } from '../../spacer';
 
-export default {
+const meta: ComponentMeta< typeof BaseControl > = {
 	title: 'Components/BaseControl',
 	component: BaseControl,
-	subcomponents: { BaseControl: BaseControl.VisualLabel },
+	argTypes: {
+		help: { control: { type: 'text' } },
+		label: { control: { type: 'text' } },
+	},
+	parameters: {
+		controls: { expanded: true },
+		docs: { source: { state: 'open' } },
+	},
 };
+export default meta;
 
-const BaseControlWithTextarea = ( { id, ...props } ) => {
+const BaseControlWithTextarea: ComponentStory< typeof BaseControl > = ( {
+	id,
+	...props
+} ) => {
 	return (
 		<BaseControl id={ id } { ...props }>
 			<textarea style={ { display: 'block' } } id={ id } />
 		</BaseControl>
 	);
 };
 
-export const Default = BaseControlWithTextarea.bind( {} );
+export const Default: ComponentStory<
+	typeof BaseControl
+> = BaseControlWithTextarea.bind( {} );
 Default.args = {
 	__nextHasNoMarginBottom: true,
-	id: 'textarea-1',
-	label: '',
-	hideLabelFromVision: false,
-	help: '',
-};
-
-export const WithLabel = BaseControlWithTextarea.bind( {} );
-WithLabel.args = {
-	...Default.args,
+	id: 'textarea-default-1',
 	label: 'Label text',
 };
 
 export const WithHelpText = BaseControlWithTextarea.bind( {} );
 WithHelpText.args = {
-	...WithLabel.args,
+	...Default.args,
+	id: 'textarea-with-help-text-1',
 	help: 'Help text adds more explanation.',
 };
 
-export const WithVisualLabel = ( { visualLabelChildren, ...props } ) => {
+/**
+ * `BaseControl.VisualLabel` is used to render a purely visual label inside a `BaseControl` component.
+ *
+ * It should only be used in cases where the children being rendered inside `BaseControl` are already accessibly labeled,
+ * e.g., a button, but we want an additional visual label for that section equivalent to the labels `BaseControl` would
+ * otherwise use if the `label` prop was passed.
+ */
+export const WithVisualLabel: ComponentStory< typeof BaseControl > = (
+	props
+) => {
+	BaseControl.VisualLabel.displayName = 'BaseControl.VisualLabel';
```

Comment:
Interestingly, when I try to set this display name in the `base-control/index.tsx` file instead of here, the docgen refuses to extract types from `BaseControl` 🤷 I don't think we'll run into this problem a lot though. Also I guess it isn't too weird that we're setting the display name here, because it's mainly to make the Storybook code snippet understandable.

Reply:
> Could this be related to the `subcomponents` bug that you mentioned before?

My hunch is that it's more related to the thing where we need to `export` the component individually as a named export (e.g. [here](https://github.com/WordPress/gutenberg/pull/38842/commits/e7eca9f742950066dc7fe23598a09692bd272471#diff-eb4d5f0199b10565b07648a8fd189fe7efb78ecc40685a6f0d10a0fd80b53bdeR34)) for the docgen to extract prop types. If you try to change a property on that exported module directly in the same file (e.g. `ConnectedHeading.displayName = 'foo'`), then the `export` suddenly stops doing its magic.

🙂 **Docgen works as expected**

```jsx
// component.tsx
export const MyComponent = /* ... */;

// consumer.tsx
import { MyComponent } from "./component";
MyComponent.displayName = "Hello";
```

☠️ **Docgen ignores my prop types**

```jsx
// component.tsx
export const MyComponent = /* ... */;
MyComponent.displayName = "Hello";
```

I'm not familiar enough with the internals to say why this happens, but to someone who is familiar, I bet the reason is quite obvious!